lib/auth/keystore/health.go

```diff
@@ -0,0 +1,102 @@
+package keystore
+
+import (
+	"context"
+	"crypto"
+	"io"
+	"log/slog"
+	"sync/atomic"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+)
+
+const (
+	failureThreshold   = 3
+	successThreshold   = 3
+	healthCheckTimeout = time.Second * 10
+)
+
+var retryInterval = time.Second * 10
+
+// passiveHealthChecker depends on client interactions with the keystore
+// to trigger failures. Once triggered the probe function will be retried
+// until it passes the success threshold.
+type passiveHealthChecker struct {
+	callback func(error)
+	busy     atomic.Bool
+	log      *slog.Logger
+	clock    clockwork.Clock
+}
+
+type probeFunc func(context.Context) error
+
+// tryProbe will call the probeFunc until the number of consecutive successful
+// calls passes the successThreshold. This is a noop if a previous probe is still\
+// running.
+func (h *passiveHealthChecker) tryProbe(ctx context.Context, probe probeFunc) {
+	if swapped := h.busy.CompareAndSwap(false, true); !swapped {
+		return
+	}
+	go h.probeUntilHealthy(ctx, probe)
+}
+
+func (h *passiveHealthChecker) probeUntilHealthy(ctx context.Context, probe probeFunc) {
+	var oks, fails uint
+	start := h.clock.Now()
+	timer := h.clock.NewTimer(retryInterval)
+	defer h.busy.Store(false)
+
+	h.log.DebugContext(ctx, "Starting passive health check probe")
+	for {
+		h.log.InfoContext(ctx, "Trying passive health check probe", "duration", h.clock.Since(start), "fails", fails, "oks", oks)
```

Comment:
> Questions
Is failing on err != nil too strict? I am worried of hitting some state that will never recover and the probe function will fail forever. for example if an auth server is retrying a sign operation for a key that gets deleted.

I wonder if there is really a need to switch to this non-passive probeUntilHealthy mode where we keep trying the signature over and over until it succeeds, or we could just passively track the pass/fails of real signature attempts

Reply:
If an auth server is failing traffic could be routed away from it, then there may not be any requests to trigger a success.